src/test/java/org/broadinstitute/hellbender/utils/read/CigarUtilsUnitTest.java

```diff
@@ -337,4 +337,173 @@ public void testComputeCigar(String s1, String s2, String expectedCigar) throws
+        final int actual = CigarUtils.countRightHardClippedBases(cigar);
+        int expected = 0;
+        for (final CigarElement element : CigarUtils.invertCigar(cigar).getCigarElements()) {
+            if (element.getOperator() != CigarOperator.H) {
+                break;
+            }
+            expected += element.getLength();
+        }
+        Assert.assertEquals(actual, expected);
+    }
+
+    @Test(dataProvider = "randomValidAndFunkyCigars")
+    public void testSoftClip(final Cigar cigar) {
+        final Cigar actual = CigarUtils.softReclip(cigar);
+        final Cigar expected = CigarUtils.combineAdjacentCigarElements(new Cigar(
+                cigar.getCigarElements().stream()
+                        .map(ce -> ce.getOperator().isClipping() ? new CigarElement(ce.getLength(), CigarOperator.SOFT_CLIP) : ce)
+                        .collect(Collectors.toList())
+        ));
+
+        Assert.assertEquals(actual, expected);
+    }
+
+    @Test(dataProvider = "randomValidAndFunkyCigars")
+    public void testReadLength(final Cigar cigar) {
+        final int actual = CigarUtils.countUnclippedReadBases(cigar);
+        final int expected = cigar.getCigarElements().stream()
+                .filter(ce -> ce.getOperator().consumesReadBases() || ce.getOperator().isClipping())
+                .mapToInt(CigarElement::getLength).sum();
+        Assert.assertEquals(actual, expected);
+    }
+
+    @Test(dataProvider = "randomValidAndFunkyCigars")
+    public void testHardClip(final Cigar cigar) {
+        final Cigar actual = CigarUtils.hardReclip(cigar);
+        final Cigar expected = CigarUtils.combineAdjacentCigarElements(new Cigar(
+                cigar.getCigarElements().stream()
+                        .map(ce -> ce.getOperator().isClipping() ? new CigarElement(ce.getLength(), CigarOperator.HARD_CLIP) : ce)
+                        .collect(Collectors.toList())
+        ));
+        Assert.assertEquals(actual, expected);
+    }
+
+    @DataProvider(name = "randomValidAndFunkyCigars")
+    public Object[][] randomValidAndFunkyCigars() {
+        final Object[][] randomValidCigars = randomValidCigars();
+        final List<Object[]> result = new ArrayList<>(randomValidCigars.length + 10);
+        result.addAll(Arrays.asList(randomValidCigars));
+        result.add(new Object[] {TextCigarCodec.decode("10S20S30S")}); // all softclips.
+        result.add(new Object[] {TextCigarCodec.decode("10H10S")}); // H and S.
+        result.add(new Object[] {TextCigarCodec.decode("10H10S30H")}); // H, S and H.
+        result.add(new Object[] {TextCigarCodec.decode( "10S30H")}); // S and H.
+        result.add(new Object[] {TextCigarCodec.decode("10H10S10I10M10D10S10H")}); // I and D operation in the extreme of the alignment.
+        return result.toArray(new Object[result.size()][]);
+    }
+
+    @DataProvider(name = "randomValidCigars")
+    public static Object[][] randomValidCigars() {
```

Comment:
I like this!

Reply:
I extracted it into ```CigarTestUtils```, it accepts a ```Random``` instance ... you need to seed it appropriately.